External liqudity incentives: Support multiple exchanges

ilya-bobyr opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/futureswap/fs-cli-v4/26)
<!-- Reviewable:end -->

Alexsey commented on src/uniswap.ts:
```diff
@@ -409,27 +465,46 @@ export const getReportOptions = (
+    mainnet: true,
+    rinkeby: true,
+  };
+  if (unsupported[lcNetwork]) {
```
With type changes above this `if` may be removed because the type system would ensure that value is correct

ilya-bobyr replied:
This was somewhat of a special case.
Explaining why `mainnet` and `rinkeby` specifically are not supported at the moment.
But, I guess, it is fine to just remove it.
As with proper types, I would actually need to pass unsupported values deeper into the application logic.

Alexsey reviewed (commented):
I have reviewed mostly the TypeScript itself part. It would be nice if someone else would take a look on the business logic changes

kienswap reviewed (changes requested):
> Reviewed 3 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/26)* status: 3 of 11 files reviewed, 2 unresolved discussions (waiting on @gerben-stavenga and @ilya-bobyr)

___
*[src/externalLiquidityIncentives.ts, line 112 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/26#-Mv6L402981Rujdtb6X5:-Mv6L403-PRROj2YFGZe:b-mk8qng) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/b5d4e56982d05038074d803b74f2c40a2ca406e4/src/externalLiquidityIncentives.ts#L112)):*
> ```XML
>         ),
>       async (argv) => {
>         const { network, pair } = uniswap.getNetworkAndPair(getNetwork, argv);
> ```

Do we have plans to support external liquidity beyond Uniswap?


<!-- Sent from Reviewable.io -->

ilya-bobyr commented on src/externalLiquidityIncentives.ts:
```diff
@@ -609,6 +611,28 @@ const removeAccountant = async (
+}) => {
+  const { from, to, incentivesTotal, noLiquidityIncentives } = distribution;
+
+  const numberFormat = new Intl.NumberFormat(undefined, {
```
Done.

ilya-bobyr commented on src/utils.ts:
```diff
@@ -1,19 +1,52 @@
+
+  if (Number.isNaN(v)) {
+    throw new Error(
+      `${context}\n` + `"${name}" is not an number, observed value: "${v}"`
```
Done.

ilya-bobyr commented on src/uniswap/incentives.ts:
```diff
@@ -191,30 +203,38 @@ export const incentivesDistribution = (
+   * tokens.  Uses `incentivesDecimals` as precision when computing fixed decimal numbers.
+   */
+  const incentivesForTokens = (tokens: bigint): number => {
+    const asFixedDecimal =
```
This is probably OK, as there seems to be no way to avoid inaccuracies when we convert 256 bit values into floating point integers with only 53 bits of mantissa.
Inaccuracies will occur first in the lowest bits, which represent digits 14 decimal places after the point.
Not sure what is an alternative here.

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -85,8 +85,8 @@ export class MintAsJson {
     public readonly l: string,
-    public readonly usdc: number,
-    public readonly weth: number,
+    public readonly a0: string,
```
JSON files that holds liquidity balances is pretty big.
5.3MB at the moment.
As these fields are repeated many times in there, their length matters.
I must admit, that I did not measure performance of this particular optimization.
But I've seen that JSON parsing can take seconds for large files.

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -453,6 +447,94 @@ export class PoolBalancesAsJson {
+        [pair: string]: PairBalances;
+      } = {};
+      for (const [pair, pairData] of Object.entries(data)) {
+        try {
```
This block adds the `pair` value into the error message.
Discussed in a call.

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -453,6 +447,94 @@ export class PoolBalancesAsJson {
+        try {
+          pairs[pair] = await PairBalances.parse(pairData);
+        } catch (err) {
+          throw new Error(`Failed to read balances for: ${pair}\n` + err);
```
> It's a little beat strange to have both template literals and strings concatenation at the same time.

The idea is to use concatenation to separate strings that span multiple lines.
This way they also span multiple lines in the source code.
The formatter puts them on the same line here.
But still, it seems easier to see that "\n" when it ends a backtick block, I think.

> More convenient would be to have it like
> ```typescript
> throw new Error(`Failed to read balances for: ${pair}\n${err}`);
> ```

Not sure why exactly it is more convenient.
I would say that it is actually more syntactic noise.
Depends on personal habits, as you may not notice those `$` and `{}`.
But, I think,

```typescript
          throw new Error(`Failed to read balances for: ${pair}\n` + err);
```

is easier to read.

> And of course the best way to write errors is with `cause`

It would be nice to pass in `err` via the `cause` mechanism.
What do you think is the best way to use it?
Add type exceptions on every use?

Alexsey replied:
TS 4.6 is going to be released in two weeks and I have tried types the nightly build - they are working as expected. Let's leave it as-is and then fix it after the release

ilya-bobyr commented on src/uniswap.ts:
```diff
@@ -46,41 +47,39 @@ export enum ReportFormat {
+
 const CONFIGURATIONS: {
-  [network: string]: Config;
+  [network: string]: {
```
Done.

ilya-bobyr commented on src/uniswap.ts:
```diff
@@ -46,41 +47,39 @@ export enum ReportFormat {
+export enum Pair {
+  ETHUSDC = "ETHUSDC",
+}
+
```
Done.

ilya-bobyr commented on src/uniswap.ts:
```diff
@@ -409,27 +465,46 @@ export const getReportOptions = (
 
-export const configForNetwork = (network: string): Config => {
+export const configForNetworkAndPair = (
+  network: string,
```
I do not see why we would not just want to make `Network` public.
I actually put it into `config/common.ts`.

ilya-bobyr commented on src/uniswap.ts:
```diff
@@ -46,41 +47,39 @@ export enum ReportFormat {
 const CONFIGURATIONS: {
-  [network: string]: Config;
+  [network: string]: {
+    [pair: string]: Config;
```
Done.

ilya-bobyr commented on src/uniswap.ts:
```diff
@@ -409,27 +465,46 @@ export const getReportOptions = (
+
+  const config = networkConfig[pair.toString()];
+
+  if (config === undefined) {
```
`CONFIGURATION` object might still be missing a required key.
The fact that `pair` has type `Pair` does not guarantee that all possible value of type `Pair` have were used as keys in `CONFIGURATION.

ilya-bobyr commented on src/binance.ts:
```diff
@@ -68,52 +74,132 @@ export class PriceStore {
-      return new PriceStore(startTime, prices);
+      return new PriceStore(pairs);
     } catch (err) {
       throw new Error(`Failed to read prices from: ${path}\n` + err);
```
This is a duplicate from `liquidity.ts`.
Let's discuss it there and, if necessary, I will make identical changes everywhere.

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -490,94 +572,78 @@ export class BalancesStore {
+     * TODO We should look into an alternative way of JSON serialization/deserialization.  Current
+     * approach is both fragile and very verbose.
+     */
+    setStrNotationRepetition(100000);
```
Done.

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -490,94 +572,78 @@ export class BalancesStore {
+    expectedPoolAddress: string
+  ) {
+    if (expectedFirstBlock != this.firstBlock) {
+      throw new Error(
```
I do not have a preference one way or another.
I think, we got this style from our Java code, where `[line, line, line].join('\n')` is not an option.

> So it's preferable to concatenate lines with `+` construct them as `[line, line, line].join('\n')` (I think we prefer `+` across our codebase, but both are ok)

I do not understand which one are you saying is preferred.
It seems that `[line, line, line].join('\n')` is a bit more regular, in the sense that lines are more a like.
While, when combining lines with the `+` operator, the first and the last lines look different.

Alexsey replied:
I meant that we are using `+` more often across the project, so probably it's better to stick with the style we already have. But I'm absolutely ok with switching to `.join` style if you like it more - let it just be our explicit decision, so we can follow it in the future code

ilya-bobyr commented on src/uniswap/liquidity.ts:
```diff
@@ -490,94 +572,78 @@ export class BalancesStore {
+    }
+
+    if (expectedPoolAddress.toLowerCase() != this.poolAddress.toLowerCase()) {
       throw new Error(
```
Done.

ilya-bobyr commented on src/binance.ts:
```diff
@@ -68,52 +74,132 @@ export class PriceStore {
+    expectedStartTime: Date
+  ) {
+    if (this.startTime.getTime() != expectedStartTime.getTime()) {
+      throw new Error(
```
Done.

ilya-bobyr commented on src/binance.ts:
```diff
@@ -68,52 +74,132 @@ export class PriceStore {
+        [pair: string]: PairPrices;
+      } = {};
+      for (const [pair, pairData] of Object.entries(data)) {
+        try {
```
As we discussed, this adds the `pair` value into the error message.
While the outer try/catch adds the `path` value.

ilya-bobyr reviewed (commented):
> Dismissed @kienswap from a discussion.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/26)* status: 0 of 13 files reviewed, 1 unresolved discussion (waiting on @Alexsey, @gerben-stavenga, and @kienswap)

___
*[src/externalLiquidityIncentives.ts, line 112 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/26#-Mv6L402981Rujdtb6X5:-MvNGnEW344mAbWxAj4o:b-yvh97c) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/b5d4e56982d05038074d803b74f2c40a2ca406e4/src/externalLiquidityIncentives.ts#L112)):*
<details><summary><i>Previously, kienswap wrote…</i></summary><blockquote>

Do we have plans to support external liquidity beyond Uniswap?
</blockquote></details>

I do not know.
If we would decide to support a different AMM to incentivize, we would have to change this code a bit.
Most of it will probably be reusable.
But the process of extracting balances would be different.
And we also rely on the Uniswap notion of "liquidity".
Unless the AMM we are adding would be based on Uniswap we would have to make additional changes.


<!-- Sent from Reviewable.io -->

ilya-bobyr commented:
@Alexsey Thank you for the review.
I think, I have addressed all the comments and marked all conversations that were addressed as resolved.
Except for two threads that you may want to reply in.
Though, maybe you want to look at all the answers, as in a few cases I just closed the conversation without making any changes.

I am not sure if I should have kept everything unresolved for you to check?
